### PR TITLE
SpanContext hides the TraceId/SpanId implementations

### PR DESCRIPTION
**Pull request by jkwatson**
This removes the usage of wrapper classes for the ids. 
Changes the SpanContext creation methods to accept `CharSequence`s and an offset for the ids.
Changes the IdGenerator interface to generate String ids.

Relevant issue: #1314
Extensive discussion about this took place in this draft PR: #1374

### CODE REVIEW

**Comment by jkwatson** [comment]
I apologize for the size of this PR. Most of it is mechanical changes in types, so hopefully review will go quickly.

File: api/src/main/java/io/opentelemetry/trace/BigendianEncoding.java
Comment (anuraaga) [inline api/src/main/java/io/opentelemetry/trace/BigendianEncoding.java:181]: It might affect other places as well, but while we're here, let's use a non-unicode check for `isDigit`

File: api/src/main/java/io/opentelemetry/trace/BigendianEncoding.java
Comment (anuraaga) [inline api/src/main/java/io/opentelemetry/trace/BigendianEncoding.java:141]:
```suggestion
  static byte[] bytesFromBase16(CharSequence value, int offset, int length) {
```

I think this pattern is very prevelant in the JDK so good to follow the idiom

File: api/src/main/java/io/opentelemetry/trace/BigendianEncoding.java
Comment (anuraaga) [inline api/src/main/java/io/opentelemetry/trace/BigendianEncoding.java:142]: Think the loop is still very easy even without the subsequence

File: api/src/main/java/io/opentelemetry/trace/SpanContext.java
Comment (anuraaga) [inline api/src/main/java/io/opentelemetry/trace/SpanContext.java:113]:
```suggestion
        /* remote=*/ remote);
```

Isn't it?

File: extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/Common.java
Comment (anuraaga) [inline extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/Common.java:56]: We can replace some of the methods in this class for checking validity I think

File: api/src/main/java/io/opentelemetry/trace/SpanContext.java
Comment (anuraaga) [inline api/src/main/java/io/opentelemetry/trace/SpanContext.java:126]:
```suggestion
  public String getTraceIdAsHexString() {
```

Our spec now uses hex too and I think it's more familiar to people than base 16
https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#retrieving-the-traceid-and-spanid

File: api/src/main/java/io/opentelemetry/trace/SpanContext.java
Comment (anuraaga) [inline api/src/main/java/io/opentelemetry/trace/SpanContext.java:101]: Since our accessors (e.g. `getTraceId()` use `String`, I don't think there's an advantage to using `CharSequence` in the factories. Can we use `String` everywhere?

File: api/src/main/java/io/opentelemetry/trace/SpanId.java
Comment (anuraaga) [inline api/src/main/java/io/opentelemetry/trace/SpanId.java:35]:
```suggestion
  private static final String INVALID = "0000000000000000";
```

File: api/src/main/java/io/opentelemetry/trace/TraceId.java
Comment (anuraaga) [inline api/src/main/java/io/opentelemetry/trace/TraceId.java:35]:
```suggestion
  private static final String INVALID = "00000000000000000000000000000000";
```

File: api/src/main/java/io/opentelemetry/trace/SpanContext.java
Comment (anuraaga) [inline api/src/main/java/io/opentelemetry/trace/SpanContext.java:109]: I don't know if the offset versions provide enough user benefit to have (not much harder to call subsequence than grok this long method signature). There doesn't seem to be a performance benefit since we always need to take the substring anyways.

If we only had two factories, than I think we could also provide two symmetric factories that accept `byte[]` and we don't need public APIs like `TraceId.toLowerBase16(byte[])`

Comment (anuraaga) [reply api/src/main/java/io/opentelemetry/trace/SpanContext.java:109]: Though realized a byte[] version of this factory would have benefit from an offset version :)

File: api/src/main/java/io/opentelemetry/trace/SpanId.java
Comment (anuraaga) [inline api/src/main/java/io/opentelemetry/trace/SpanId.java:69]: Any thoughts on moving these to reflect they're only used for ID generation, maybe `IdGeneratorUtil`, or we could make `IdGenerator` an abstract class with protected helper methods for generating IDs from longs?

File: api/src/main/java/io/opentelemetry/trace/TraceId.java
Comment (anuraaga) [inline api/src/main/java/io/opentelemetry/trace/TraceId.java:166]:
```suggestion
  public static long getRandomPart(CharSequence traceId) {
```

We don't repeat the classname in the others, don't think we need to here either.

Comment (anuraaga) [reply api/src/main/java/io/opentelemetry/trace/TraceId.java:166]: Though this seems like it doesn't need to be public API, it even documents itself as being used in one specific sampler?

File: extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/B3PropagatorInjectorSingleHeader.java
Comment (anuraaga) [inline extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/B3PropagatorInjectorSingleHeader.java:54]:
```suggestion
    traceId.getChars(0, traceId.length(), chars, 0);
```

Similar throughout

File: api/src/main/java/io/opentelemetry/trace/propagation/HttpTraceContext.java
Comment (anuraaga) [inline api/src/main/java/io/opentelemetry/trace/propagation/HttpTraceContext.java:98]: Looks like a lot can be updated to use `String` / getChars instead of `CharSequence`

**Review by anuraaga — COMMENTED** [review]
Thanks!

File: api/src/main/java/io/opentelemetry/trace/SpanContext.java
Comment (jkwatson) [reply api/src/main/java/io/opentelemetry/trace/SpanContext.java:113]: you would think that, but it seems to shift back and forth in ways I can't predict. If you get it wrong, checkstyle/errorprone complains.

File: api/src/main/java/io/opentelemetry/trace/TraceId.java
Comment (jkwatson) [reply api/src/main/java/io/opentelemetry/trace/TraceId.java:166]: This was the method name before, so I kept it as-is.  And, it's public so any Sampler could get access to it, I believe.

File: api/src/main/java/io/opentelemetry/trace/propagation/HttpTraceContext.java
Comment (jkwatson) [reply api/src/main/java/io/opentelemetry/trace/propagation/HttpTraceContext.java:98]: yeah, it's true. I hadn't tracked them all down. I don't think it's actually harmful to declare these as CharSequence, so I wasn't aggressive at finding them all.

File: api/src/main/java/io/opentelemetry/trace/BigendianEncoding.java
Comment (jkwatson) [reply api/src/main/java/io/opentelemetry/trace/BigendianEncoding.java:142]: definitely. I thought I had caught all these, but I missed this one. will fix.

File: api/src/main/java/io/opentelemetry/trace/BigendianEncoding.java
Comment (jkwatson) [reply api/src/main/java/io/opentelemetry/trace/BigendianEncoding.java:181]: happy to fix it. Are there non-ascii digits out there?

File: api/src/main/java/io/opentelemetry/trace/SpanContext.java
Comment (jkwatson) [reply api/src/main/java/io/opentelemetry/trace/SpanContext.java:101]: My thinking (and it wasn't very strong thinking) was that to be more accepting of inputs, and more restrictive on outputs. But, I definitely don't feel too strongly about that.

File: api/src/main/java/io/opentelemetry/trace/SpanContext.java
Comment (jkwatson) [reply api/src/main/java/io/opentelemetry/trace/SpanContext.java:126]: I'll go through and change it.

File: api/src/main/java/io/opentelemetry/trace/SpanId.java
Comment (jkwatson) [reply api/src/main/java/io/opentelemetry/trace/SpanId.java:69]: I liked having all the transformations "close to home" in the relevant helpers, but it wouldn't be hard to convince me to move them. Let's do that as a follow-up, if that's ok.

File: extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/Common.java
Comment (jkwatson) [reply extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/Common.java:56]: You mean use the ones on TraceId/SpanId?

File: api/src/main/java/io/opentelemetry/trace/SpanContext.java
Comment (jkwatson) [reply api/src/main/java/io/opentelemetry/trace/SpanContext.java:109]: This was a method signature that @bogdandrutu had requested. I don't know that it's necessary at the moment, though.

File: api/src/main/java/io/opentelemetry/trace/SpanId.java
Comment (jkwatson) [reply api/src/main/java/io/opentelemetry/trace/SpanId.java:69]: Also, there are quite a few usages in test code at the moment, so I think I'd want to hold off for a follow-up PR, for sure.

File: extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/Common.java
Comment (anuraaga) [reply extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/Common.java:56]: Yeah

Comment (jkwatson) [reply extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/Common.java:56]: done

File: api/src/main/java/io/opentelemetry/trace/BigendianEncoding.java
Comment (jkwatson) [reply api/src/main/java/io/opentelemetry/trace/BigendianEncoding.java]: done

File: api/src/main/java/io/opentelemetry/trace/BigendianEncoding.java
Comment (jkwatson) [reply api/src/main/java/io/opentelemetry/trace/BigendianEncoding.java]: done

File: api/src/main/java/io/opentelemetry/trace/SpanContext.java
Comment (anuraaga) [reply api/src/main/java/io/opentelemetry/trace/SpanContext.java:101]: Got it - I'm ok with either, but have a small feeling that `String` feels more familiar to developers

File: api/src/main/java/io/opentelemetry/trace/propagation/HttpTraceContext.java
Comment (jkwatson) [reply api/src/main/java/io/opentelemetry/trace/propagation/HttpTraceContext.java:98]: I'm going to go through and get rid of the CharSequence usage and go to strictly String based API. We can always loosen it up later, if we just support String for now.

File: api/src/main/java/io/opentelemetry/trace/SpanContext.java
Comment (jkwatson) [reply api/src/main/java/io/opentelemetry/trace/SpanContext.java]: I think we should get rid of the offsets for now; we can always add another option later; we can't take it back if we leave it in.

File: api/src/main/java/io/opentelemetry/trace/SpanContext.java
Comment (jkwatson) [reply api/src/main/java/io/opentelemetry/trace/SpanContext.java]: I think occurs when you use the `@Memoized` annotation, due to the way that generates the underlying implementations.

File: api/src/main/java/io/opentelemetry/trace/SpanContext.java
Comment (jkwatson) [reply api/src/main/java/io/opentelemetry/trace/SpanContext.java]: I got rid of them. Smaller API is better for now.

File: api/src/main/java/io/opentelemetry/trace/propagation/HttpTraceContext.java
Comment (jkwatson) [reply api/src/main/java/io/opentelemetry/trace/propagation/HttpTraceContext.java]: done

**Comment by jkwatson** [comment]
@anuraaga I think I got all of your comments addressed. Thanks so much for taking the time to give it a thorough review!

File: api/src/main/java/io/opentelemetry/trace/BigendianEncoding.java
Comment (anuraaga) [inline api/src/main/java/io/opentelemetry/trace/BigendianEncoding.java:180]: I'm pretty sure they're the same, and maybe it's just me but I find writing the same code as the comment would be more readable 

```suggestion
      if (!(b >= 48 && b <= 57) || !(b >= 97 && b <= 102)) {
```

File: api/src/test/java/io/opentelemetry/trace/TraceIdTest.java
Comment (anuraaga) [inline api/src/test/java/io/opentelemetry/trace/TraceIdTest.java:85]: Think we can remove `.toString()` in this file

File: exporters/logging/src/test/java/io/opentelemetry/exporters/logging/LoggingSpanExporterTest.java
Comment (anuraaga) [inline exporters/logging/src/test/java/io/opentelemetry/exporters/logging/LoggingSpanExporterTest.java:63]: A note for potentially following up with moving `fromLongs` to ID generation, there isn't a great reason for unit tests to use the long version vs string version, presumably assertion messages matching the input is an improvement.

File: exporters/zipkin/src/main/java/io/opentelemetry/exporters/zipkin/ZipkinSpanExporter.java
Comment (anuraaga) [inline exporters/zipkin/src/main/java/io/opentelemetry/exporters/zipkin/ZipkinSpanExporter.java:136]: Can remove `toString`

File: extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/B3PropagatorInjectorMultipleHeaders.java
Comment (anuraaga) [inline extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/B3PropagatorInjectorMultipleHeaders.java:43]:
```suggestion
    setter.set(carrier, B3Propagator.SPAN_ID_HEADER, spanContext.getSpanIdAsHexString());
```

File: extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/JaegerPropagator.java
Comment (anuraaga) [inline extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/JaegerPropagator.java:206]: For another PR, we might consider moving the validity checks into `createFromRemoteParent` so propagators don't have to worry about it.

File: extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/OtTracerPropagator.java
Comment (anuraaga) [inline extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/OtTracerPropagator.java:77]:
```suggestion
    setter.set(carrier, TRACE_ID_HEADER, spanContext.getTraceIdAsHexString());
```

File: extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/OtTracerPropagator.java
Comment (anuraaga) [inline extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/OtTracerPropagator.java:78]:
```suggestion
    setter.set(carrier, SPAN_ID_HEADER, spanContext.getSpanIdAsHexString());
```

File: opentracing_shim/src/test/java/io/opentelemetry/opentracingshim/testbed/TestUtils.java
Comment (anuraaga) [inline opentracing_shim/src/test/java/io/opentelemetry/opentracingshim/testbed/TestUtils.java:183]: Don't need `toString()` (presumably it's a better assertion without it too). Next few files have it

**Review by anuraaga — APPROVED** [review]
Just nits, thanks!

File: api/src/main/java/io/opentelemetry/trace/BigendianEncoding.java
Comment (jkwatson) [reply api/src/main/java/io/opentelemetry/trace/BigendianEncoding.java:180]: heh. I agree with you. However, what you pasted isn't correct (IDEA says it's always true!). I'll figure out hot to make it what you intended. :)

File: extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/JaegerPropagator.java
Comment (jkwatson) [reply extensions/trace_propagators/src/main/java/io/opentelemetry/extensions/trace/propagation/JaegerPropagator.java:206]: That seems reasonable!

**Comment by jkwatson** [comment]
ok, all nits have been de-nitted!

File: api/src/main/java/io/opentelemetry/trace/BigendianEncoding.java
Comment (jkwatson) [reply api/src/main/java/io/opentelemetry/trace/BigendianEncoding.java]: check out the update. I think it's now much easier to understand.

**Review by anuraaga — APPROVED** [review]
LGTM thanks

**Comment by jkwatson** [comment]
I'm going to hold off a few days to merge this, both to delay to after the 0.8.0 release, and to give others (@carlosalberto and @bogdandrutu) to chime in with any issue they might have, as this is a significant API change.

File: api/src/main/java/io/opentelemetry/trace/SpanContext.java
Comment (tylerbenson) [inline api/src/main/java/io/opentelemetry/trace/SpanContext.java:90]: return `CharSequence` instead?

File: api/src/main/java/io/opentelemetry/trace/SpanId.java
Comment (bogdandrutu) [inline api/src/main/java/io/opentelemetry/trace/SpanId.java:54]: Hex? If you used hex in SpanContext method names.

File: api/src/main/java/io/opentelemetry/trace/SpanId.java
Comment (bogdandrutu) [inline api/src/main/java/io/opentelemetry/trace/SpanId.java:75]: s/getBuffer/getTemporaryBuffer?

File: api/src/main/java/io/opentelemetry/trace/SpanId.java
Comment (bogdandrutu) [inline api/src/main/java/io/opentelemetry/trace/SpanId.java:96]: Same, use Hex if you decided to use that in the SC

File: api/src/main/java/io/opentelemetry/trace/SpanContext.java
Comment (bogdandrutu) [inline api/src/main/java/io/opentelemetry/trace/SpanContext.java:63]: name the params traceIdHex, spanIdHex

File: api/src/main/java/io/opentelemetry/trace/SpanId.java
Comment (jkwatson) [reply api/src/main/java/io/opentelemetry/trace/SpanId.java:54]: yes! should fix this. good catch.

File: api/src/main/java/io/opentelemetry/trace/SpanId.java
Comment (jkwatson) [reply api/src/main/java/io/opentelemetry/trace/SpanId.java:96]: agreed. 👍

File: api/src/main/java/io/opentelemetry/trace/SpanContext.java
Comment (jkwatson) [reply api/src/main/java/io/opentelemetry/trace/SpanContext.java:90]: I think we discussed this, and decided to keep it as String for now. We can always open it up to CharSequence later, since it's a broader interface, I think.

File: api/src/main/java/io/opentelemetry/trace/SpanId.java
Comment (jkwatson) [reply api/src/main/java/io/opentelemetry/trace/SpanId.java]: doen

File: api/src/main/java/io/opentelemetry/trace/SpanId.java
Comment (jkwatson) [reply api/src/main/java/io/opentelemetry/trace/SpanId.java]: done

File: api/src/main/java/io/opentelemetry/trace/SpanId.java
Comment (jkwatson) [reply api/src/main/java/io/opentelemetry/trace/SpanId.java]: done

File: api/src/main/java/io/opentelemetry/trace/SpanContext.java
Comment (jkwatson) [reply api/src/main/java/io/opentelemetry/trace/SpanContext.java]: done

File: api/src/main/java/io/opentelemetry/trace/SpanContext.java
Comment (iNikem) [reply api/src/main/java/io/opentelemetry/trace/SpanContext.java]: You can change return types that easily. If clients expect `String` they will break on `CharSequence`. You can widen input, not output

File: api/src/main/java/io/opentelemetry/trace/SpanContext.java
Comment (jkwatson) [reply api/src/main/java/io/opentelemetry/trace/SpanContext.java]: switched all to String now. :)

File: api/src/main/java/io/opentelemetry/trace/SpanContext.java
Comment (jkwatson) [reply api/src/main/java/io/opentelemetry/trace/SpanContext.java]: true, but this isn't a public method!

**Review by carlosalberto — APPROVED** [review]

**Comment by carlosalberto** [comment]
Amazing job @jkwatson sorry for the late review!

**Comment by anuraaga** [comment]
Awesome!

**Pull request merged**
